src/backend/cuda/kernel/hamming.hpp

```diff
@@ -0,0 +1,450 @@
+#include <memory.hpp>
+#include <platform.hpp>
+
+namespace cuda
+{
+
+namespace kernel
+{
+
+static const unsigned THREADS = 256;
+
+template<typename T, unsigned feat_len, bool use_shmem>
+__global__ void hamming_matcher_unroll(
+    unsigned* out_idx,
+    unsigned* out_dist,
+    CParam<T> query,
+    CParam<T> train,
+    const unsigned max_dist)
+{
+    unsigned nquery = query.dims[0];
+    unsigned ntrain = train.dims[0];
+
+    unsigned f = blockDim.x * blockIdx.x + threadIdx.x;
+    unsigned tid = threadIdx.x;
+
+    __shared__ unsigned s_dist[THREADS];
+    __shared__ unsigned s_idx[THREADS];
+
+    extern __shared__ char smem[];
+    T* s_query = (T*)smem;
+    T* s_train = (T*)smem + feat_len;
+
+    s_dist[tid] = max_dist;
+    s_idx[tid]  = 0xffffffff;
+
+    bool valid_feat = (f < ntrain);
+
+    if (valid_feat) {
+        // Copy blockDim.x training features to shared memory
+        if (use_shmem) {
+            #pragma unroll
+            for (unsigned i = 0; i < feat_len; i++) {
+                s_train[i * blockDim.x + tid] = train.ptr[i * ntrain + f];
+            }
+        }
+    }
+    __syncthreads();
+
+    for (unsigned j = 0; j < nquery; j++) {
+        s_dist[tid] = max_dist;
+
+        // Load one query feature that will be tested against all training
+        // features in current block
+        if (tid < feat_len && valid_feat) {
+            s_query[tid] = query.ptr[tid * nquery + j];
+        }
+        __syncthreads();
+
+        unsigned dist = 0;
+        if (valid_feat) {
+            #pragma unroll
+            for (unsigned k = 0; k < feat_len; k++) {
+                // Calculate Hamming distance for 32-bits of descriptor and
+                // accumulates to dist
+                if (use_shmem) {
+                    dist += __popc(s_train[k * blockDim.x + tid] ^ s_query[k]);
+                }
+                else {
+                    dist += __popc(train.ptr[k * ntrain + f] ^ s_query[k]);
+                }
+            }
+
+            // Only stores the feature index and distance if it's smaller
+            // than the best match found so far
+            s_dist[tid] = dist;
+            s_idx[tid]  = f;
+        }
+        __syncthreads();
+
+        // Find best match in training features from block to the current
+        // query feature
+        if (tid < 128) {
+            if (s_dist[tid + 128] < s_dist[tid]) {
+                s_dist[tid] = s_dist[tid + 128];
+                s_idx[tid]  = s_idx[tid + 128];
+            }
+        }
+        __syncthreads();
+        if (tid < 64) {
+            if (s_dist[tid + 64] < s_dist[tid]) {
+                s_dist[tid] = s_dist[tid + 64];
+                s_idx[tid]  = s_idx[tid + 64];
+            }
+        }
+        __syncthreads();
+        if (tid < 32) {
+            if (s_dist[tid + 32] < s_dist[tid]) {
+                s_dist[tid] = s_dist[tid + 32];
+                s_idx[tid]  = s_idx[tid + 32];
+            }
```

Comment:
For whatever reason this is causing race conditions in newer versions of cuda. Just use __syncthreads() for every block. They are not as costly on newer hardware.